src/dwarf_parser.cpp

```diff
@@ -0,0 +1,171 @@
+#include "dwarf_parser.h"
+
+#ifdef HAVE_LIBDW
+
+#include "log.h"
+#include "utils.h"
+#include <dwarf.h>
+#include <elfutils/libdw.h>
+
+#ifdef HAVE_BCC_ELF_FOREACH_SYM
+#include <bcc/bcc_elf.h>
+#include <bcc/bcc_syms.h>
+#endif
+
+namespace bpftrace {
+
+struct FuncInfo
+{
+  std::string name;
+  Dwarf_Die die;
+};
+
+Dwarf::Dwarf(const std::string &file_path) : file_path(file_path)
+{
+  Dwfl_Callbacks callbacks = {
+    .find_debuginfo = dwfl_standard_find_debuginfo,
+    .section_address = dwfl_offline_section_address
+  };
+  dwfl = dwfl_begin(&callbacks);
+  if (dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1) ==
+      NULL)
+  {
+    LOG(WARNING) << "Could not read DWARF info from " + file_path;
+  }
+  dwfl_report_end(dwfl, NULL, NULL);
+
+  Dwarf_Addr bias;
+  has_data_ = dwfl_nextcu(dwfl, NULL, &bias) != NULL;
+}
+
+Dwarf::~Dwarf()
+{
+  dwfl_end(dwfl);
+}
+
+int get_func_die_cb(Dwarf_Die *func_die, void *arg)
+{
+  auto *func_info = static_cast<struct FuncInfo *>(arg);
+  if (dwarf_diename(func_die) == func_info->name)
+  {
+    func_info->die = *func_die;
+    return DWARF_CB_ABORT;
+  }
+  return DWARF_CB_OK;
+}
+
+std::optional<Dwarf_Die> Dwarf::get_func_die(const std::string &function) const
+{
+  struct FuncInfo func_info = { .name = function, .die = {} };
+
+  Dwarf_Die *cudie = nullptr;
+  Dwarf_Addr cubias;
+  while ((cudie = dwfl_nextcu(dwfl, cudie, &cubias)) != nullptr)
+  {
+    if (dwarf_getfuncs(cudie, get_func_die_cb, &func_info, 0) != 0)
+      return func_info.die;
+  }
+
+  return std::nullopt;
+}
+
+Dwarf_Die type_of(Dwarf_Die &die)
+{
+  Dwarf_Attribute attr;
+  Dwarf_Die type_die;
+  dwarf_formref_die(dwarf_attr_integrate(&die, DW_AT_type, &attr), &type_die);
+  return type_die;
+}
+
+std::map<std::string, Dwarf_Die> Dwarf::function_param_dies(
```

Comment:
If you don't need ordering, prefer `std::unordered_map` -- it'll be faster. And I don't think we need ordering here.